src/app/SimpleSubscriptionResumptionStorage.cpp

```diff
@@ -0,0 +1,589 @@
+    CHIP_ERROR err = FindByScopedNodeId(subscriptionNode, subscriptions, 1); // ask to allocate 1 extra space for new subscription
+    if (err != CHIP_NO_ERROR)
+    {
+        return err;
+    }
+
+    // if this is the first subscription, allocate space for 1
+    if (!subscriptions.mSize)
+    {
+        subscriptions.mSubscriptions = std::unique_ptr<SubscriptionInfo[]>(new (std::nothrow) SubscriptionInfo[1]);
+        if (!subscriptions.mSubscriptions)
+        {
+            return CHIP_ERROR_NO_MEMORY;
+        }
+    }
+
+    // Sanity check for duplicate subscription and remove
+    for (size_t i = 0; i < subscriptions.mSize; i++)
+    {
+        if (subscriptionInfo.mSubscriptionId == subscriptions[i].mSubscriptionId)
+        {
+            subscriptions.mSize--;
+            // if not last element, move last element here, essentially deleting this
+            if (i < subscriptions.mSize)
+            {
+                subscriptions[i] = std::move(subscriptions[subscriptions.mSize]);
+            }
+            break;
+        }
+    }
+
+    // Sanity check this will not go over fabric limit - count
+    size_t totalSubscriptions = subscriptions.mSize;
+    for (size_t i = 0; i < subscriptionIndex.mSize; i++)
+    {
+        // This node has already been loaded and counted
+        if (subscriptionNode == subscriptionIndex[i])
+        {
+            continue;
+        }
+
+        SubscriptionList otherSubscriptions;
+        err = FindByScopedNodeId(subscriptionIndex[i], otherSubscriptions);
+        if (err == CHIP_NO_ERROR)
+        {
+            totalSubscriptions += otherSubscriptions.mSize;
+        }
+    }
+    // Not using CHIP_IM_MAX_NUM_SUBSCRIPTIONS_PER_FABRIC per explanation
+    // in MaxStateSize() header comment block
+    if (totalSubscriptions == CHIP_IM_MAX_NUM_SUBSCRIPTIONS)
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    // Merge new subscription in and save
+    subscriptions[subscriptions.mSize++] = std::move(subscriptionInfo);
+    return SaveSubscriptions(subscriptionNode, subscriptions);
+}
+
+CHIP_ERROR SimpleSubscriptionResumptionStorage::Delete(const SubscriptionInfo & subscriptionInfo)
+{
+    ScopedNodeId subscriptionNode = ScopedNodeId(subscriptionInfo.mNodeId, subscriptionInfo.mFabricIndex);
+    // load existing subscriptions, then search for subscription
+    SubscriptionList subscriptions;
+    CHIP_ERROR err = FindByScopedNodeId(subscriptionNode, subscriptions);
+
+    if (err != CHIP_NO_ERROR)
+    {
+        return err;
+    }
+
+    bool subscriptionsChanged = false;
+    for (size_t i = 0; i < subscriptions.mSize; i++)
+    {
+        if (subscriptionInfo.mSubscriptionId == subscriptions[i].mSubscriptionId)
+        {
+            subscriptions.mSize--;
+            // if not last element, move last element here, essentially deleting this
+            if (i < subscriptions.mSize)
+            {
+                subscriptions[i] = std::move(subscriptions[subscriptions.mSize]);
```

Comment:
Are you sure `std::swap` is not what you want here?

Reply:
I've since revamped the storage scheme as Michael suggested and this is no longer there